mongo/src/main/java/com/redhat/lightblue/mongo/crud/BulkUpdateCtx.java

```diff
@@ -0,0 +1,185 @@
+        }
+    }
+    
+    BulkWriteOperation bwo;
+    /**
+     * The new docver value
+     */
+    String newDocver;
+    /**
+     * Map of _id -> docver
+     */
+    Map<Object,String> docIdDocVerMap;
+    List<UpdateDoc> updateDocs;
+    int numUpdated;
+    int numErrors;
+
+    private final DBCollection collection;
+
+    public BulkUpdateCtx(DBCollection collection) {
+        this.collection=collection;
+        reset();
+    }
+    
+    private void reset() {
+        bwo=collection.initializeUnorderedBulkOperation();
+        newDocver=ObjectId.get().toString();
+        docIdDocVerMap=new HashMap<>();
+        updateDocs=new ArrayList<>();
+    }
+    
+    public void addDoc(UpdateDoc doc,DBObject replaceDoc) {
+        bwo.find(new BasicDBObject("_id", doc.id).
+                 append(Translator.DOC_VERSION_FULLPATH_STR,doc.docver)).
+            replaceOne(replaceDoc);
+        updateDocs.add(doc);
+        docIdDocVerMap.put(doc.id,doc.docver);
+        doc.docCtx.setCRUDOperationPerformed(CRUDOperation.UPDATE);
+        doc.docCtx.setUpdatedDocument(doc.docCtx);
+    }
+    
+    private void checkConcurrentModifications(DBCollection collection) {
+        Set<Object> failedIds=Utils.checkFailedUpdates(collection,newDocver,docIdDocVerMap);
+        for(Object id:failedIds) {
+            for(UpdateDoc doc:updateDocs) {
+                if(doc.id.equals(id)&&!doc.docCtx.hasErrors()) {
+                    doc.docCtx.addError(Error.get("update",MongoCrudConstants.ERR_CONCURRENT_MODIFICATION,doc.id.toString()));
+                    break;
+                }
+            }
+        }
+    }
+    
+    public void execute(WriteConcern writeConcern) {
+        try {
+            BulkWriteResult result;
+            if (writeConcern == null) {
+                LOGGER.debug("Bulk updating docs");
+                result=bwo.execute();
+            }
+            else {
+                LOGGER.debug("Bulk updating docs with writeConcern={} from execution", writeConcern);
+                result=bwo.execute(writeConcern);
+            }
+            numUpdated+=result.getModifiedCount();
+            if(result.getMatchedCount()<updateDocs.size()) {
+                // There are docs that didn't match, concurrent modification
+                checkConcurrentModifications(collection);
+            }
+        } catch (BulkWriteException bwe) {
+            LOGGER.error("Bulk write exception", bwe);
+            handleBulkWriteError(bwe.getWriteErrors(),"update",(int i)->updateDocs.get(i).docCtx);
+            checkConcurrentModifications(collection);
+            numUpdated+=bwe.getWriteResult().getModifiedCount();
+        }
+        for(UpdateDoc doc:updateDocs)
+            if(doc.docCtx.hasErrors())
+                numErrors++;
+        reset();
+    }
+
+    public interface GetCB {
+        DocCtx getCtx(int index);
+    }
+    
+    public static <T> void handleBulkWriteError(List<BulkWriteError> errors,String op,GetCB cb) {
```

Comment:
Cleaned.

On Tue, Aug 9, 2016 at 2:04 PM, Alec Henninger <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/BulkUpdateCtx.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/299#discussion_r74133801>
> :
>
> > +            LOGGER.error("Bulk write exception", bwe);
> > +            handleBulkWriteError(bwe.getWriteErrors(),"update",(int i)->updateDocs.get(i).docCtx);
> > +            checkConcurrentModifications(collection);
> > +            numUpdated+=bwe.getWriteResult().getModifiedCount();
> > +        }
> > +        for(UpdateDoc doc:updateDocs)
> > +            if(doc.docCtx.hasErrors())
> > +                numErrors++;
> > +        reset();
> > +    }
> > +
> > +    public interface GetCB {
> > +        DocCtx getCtx(int index);
> > +    }
> > +
> > +    public static <T> void handleBulkWriteError(List<BulkWriteError> errors,String op,GetCB cb) {
>
> nit: unnecessary generic type declaration
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/299/files/4445c9305f6b8f7a6c756db712b5874118e095f2#r74133801>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDXQ2hNdBzdaAJV09oLQlzeGRADgsks5qeN1GgaJpZM4JfOdd>
> .
>